stdpopsim/genomes.py

```diff
@@ -352,20 +369,49 @@ def species_contig(
                 raise ValueError("Cannot specify sequence length for named contig")
             if inclusion_mask is not None and exclusion_mask is not None:
                 raise ValueError("Cannot specify both inclusion and exclusion masks")
+
             chrom = species.genome.get_chromosome(chromosome)
+            if left is None:
+                left = 0
+            else:
+                if not 0 <= left < chrom.length:
+                    raise ValueError(
+                        f"Left coordinate {left} falls outside chromosome "
+                        f"{chrom.id} with length {chrom.length}."
+                    )
+            if right is None:
+                right = chrom.length
+            else:
+                if not left < right <= chrom.length:
+                    raise ValueError(
+                        f"Right coordinate {right} falls outside chromosome "
+                        f"{chrom.id} with length {chrom.length}."
+                    )
+
             if genetic_map is None:
                 logger.debug(f"Making flat chromosome {length_multiplier} * {chrom.id}")
                 gm = None
+                # TODO: is there a need for `length_multiplier`, given the ability
+                # to specify an arbitary interval? Combining both here results in
+                # a confusing coordinate system.
                 recomb_map = msprime.RateMap.uniform(
-                    round(chrom.length * length_multiplier), chrom.recombination_rate
+                    round((right - left) * length_multiplier), chrom.recombination_rate
                 )
+                left = round(left * length_multiplier)
+                right = left + int(recomb_map.sequence_length)
             else:
                 if length_multiplier != 1:
                     raise ValueError("Cannot use length multiplier with empirical maps")
                 logger.debug(f"Getting map for {chrom.id} from {genetic_map}")
                 gm = species.get_genetic_map(genetic_map)
                 recomb_map = gm.get_chromosome_map(chrom.id)
+                recomb_map = recomb_map.slice(left=left, right=right, trim=True)
 
+            # TODO: previously, a mask with intervals falling outside the
+            # contig would throw an error in tskit. Now, the mask is silently
+            # clipped to the contig boundaries.  Should a warning be raised if
+            # this happens? (A warning will be raised if ALL masked intervals
+            # are outside of the contig).
```

Comment:
Is there a reason not to just explicitly raise an error in this case?

Reply:
Hmm, say there's an available chromosome-wide mask and annotation, and the user wants to use these whilst simulating some small region. One option is that the user clips the mask/annotation to the region themselves, and stdpopsim raises errors if these don't all fall within the region. Another option is that stdpopsim does the clipping and warns if something extreme happens, like all mask/annotation intervals falling outside of the region (this is the current strategy in the PR). I could go either way -- not sure what the right trade-off is between ease-of-use and preventing miss-specification.

Another possibility is to go the latter route (stdpopsim does the clipping-to-region) but first check that the mask/annotation matches the chromosome boundaries and raise an error if not.

Reply:
Ok, I get it. In that case I think you've made the right decision.